### PR TITLE
HUD interface improvements

### PR DESCRIPTION
**Pull request by stechio**
This PR  aims to enhance the visual feedback of the HUD interface:

- target display:
  - [ADD] **target hit detector** (ring "target hit"): _highlights whenever the target takes damage_ (rendered as a reddish thick inner ring for 200 milliseconds).
&nbsp;
endless-sky_feat_hud-refine_target

- flagship display:
  - [ADD] **flagship hit detector** (ring "hit"): _highlights whenever the flagship takes damage_ (rendered as a reddish thick inner ring for 200 milliseconds).
  - [ADD] **target bearing** (point "player target"): backs up the bearing currently on target display for convenience (_this way, the user can easily collimate the flagship direction with the target location_).
&nbsp;
endless-sky_feat_hud-refine_flagship

### CODE REVIEW

File: source/Engine.cpp [hunk 1]
```diff
@@ -568,7 +568,7 @@ void Engine::Step(bool isActive)
 	if(flagship && flagship->Hull())
 	{
 		Point shipFacingUnit(0., -1.);
-		if(Preferences::Has("Rotate flagship in HUD"))
+		if(Preferences::Has("Rotate flagship in HUD") || zoom < 1)
```
[the remark this answers is not in the document]
Comment (stechio) [reply source/Engine.cpp]: When I was trying the game (unaware of that rotation preference), I felt the fixed flagship outline quite awkward (why locking it by default while letting the target outline freely rotate at the same time? what's the UX rationale behind it?), especially considering that:

- the **viewport's coordinate system is fixed** (_in the scene it's the flagship to rotate its axis, NOT the world to pivot around its center_ while keeping the flagship's axis upright)
- **zooming out, the smaller spaceships become difficult to discern**, so _activating a visual hint of their orientation makes sense_, IMO.

Can you explain me what's the paramount reason why the flagship outline is (counterintuitively) fixed by default? Wouldn't be better to let it rotate?

[the remark this answers is not in the document]
Comment (jafdy) [reply source/Engine.cpp]: If an option exists then you never force it one way or the other, you could change the default or remove the option completely but you do not override the player decision.

[the remark this answers is not in the document]
Comment (stechio) [reply source/Engine.cpp]: ok, change cancelled.

**Comment by Bladewood** [comment]
It appears to me that this always renders at full power, for any duration, and for any amount of damage. I think it would look pretty silly to a captain in a Quarg-level ship if the alarms were to blare when a laser sparrow tries to not-even-barely scratch it. (They do blare when any hostile enters the system, but that's beside the point.)
How about making it more dynamic? To be precise, there's two things I'd like to add (which might essentially be the same):

1. Have the alpha level of the circle depend on how much damage the ship is taking, relative to its total health. (So, a few low-powered lasers would have little to no influence, whereas a skylance would be very noticeable.)
2. Have the alpha level depend on how much damage the ship is taking over time. (So, a projectile weapon will immediately highlight, whereas a beam with equal DPS will create the same highlight more gradually.)

I'm also not a fan of the vibrant red.

**Comment by stechio** [comment]
> It appears to me that this always renders at full power, for any duration, and for any amount of damage. I think it would look pretty silly to a captain in a Quarg-level ship if the alarms were to blare when a laser sparrow tries to not-even-barely scratch it. (They do blare when any hostile enters the system, but that's beside the point.)
> How about making it more dynamic? To be precise, there's two things I'd like to add (which might essentially be the same):
> 
>   1. Have the alpha level of the circle depend on how much damage the ship is taking, relative to its total health. (So, a few low-powered lasers would have little to no influence, whereas a skylance would be very noticeable. This doesn't really work as stated because a skylance actually only does 50/40 damage per frame, but I think it'd be neat in combination with the next point.)

Good point.

>   2. Have the alpha level depend on how much damage the ship is taking over time. (So, a projectile weapon will immediately highlight, whereas a beam with equal DPS will create the same highlight more gradually.)

So, we'll have to keep track of 2 components:

1. cumulative damage over last second (for low-density hit events, like beams)
2. instant damage (for high-density hit events, like projectiles)

There could be a rolling buffer of taken damage keeping track of the last second's hit events (one value for each time frame, weighted), allowing a continuously updated cumuiation into a `hitAmount`: if the action is low-density yet persistent like a beam (1 DPF, 60 DPS), it will reach the highest highlighting after 60 frames.

The actual damage represented by the detector (prior to ship health weighting) will be made of the instant damage AND the cumulative damage (so when a 60 DPF projectile impacts while a 1 DPF beam has been acting for over a second, the resulting gross damage detected will be 120).

File: source/Engine.cpp [hunk 2]
```diff
@@ -683,7 +687,11 @@ void Engine::Step(bool isActive)
 			info.SetBar("target shields", target->Shields());
 			info.SetBar("target hull", target->Hull(), 20.);
 			info.SetBar("target disabled hull", min(target->Hull(), target->DisabledHull()), 20.);
-		
+			if(targetHit)
```
Comment (Fzzr) [inline source/Engine.cpp]: Single-line loop or conditional bodies don't need braces per [style guide](http://endless-sky.github.io/styleguide/styleguide.xml). (Search for "single-line loop bodies".)

File: source/Engine.cpp [hunk 3]
```diff
@@ -592,6 +592,10 @@ void Engine::Step(bool isActive)
 		info.SetBar("shields", flagship->Shields());
 		info.SetBar("hull", flagship->Hull(), 20.);
 		info.SetBar("disabled hull", min(flagship->Hull(), flagship->DisabledHull()), 20.);
+		if(flagshipHit)
```
Comment (Fzzr) [inline source/Engine.cpp]: Single-line loop or conditional bodies don't need braces per [style guide](http://endless-sky.github.io/styleguide/styleguide.xml). (Search for "single-line loop bodies".)

File: source/Engine.cpp [hunk 4]
```diff
@@ -1762,6 +1776,21 @@ void Engine::DoCollisions(Projectile &projectile)
 			int eventType = hit->TakeDamage(projectile);
 			if(eventType)
 				eventQueue.emplace_back(gov, hit, eventType);
+
+			shared_ptr<const Ship> flagship = player.FlagshipPtr();
+			if(flagship)
+			{
+				if(flagship == hit)
```
Comment (Fzzr) [inline source/Engine.cpp]: Don't do the no-braces style here though, because there's also an `else`.

**Comment by stechio** [comment]
[Amendment 1](https://github.com/endless-sky/endless-sky/pull/4502/commits/cdfc01609cdab7357ee2db7fa311ffbedcec3532) adopts the suggestions of @Bladewood as defined in [my analysis](https://github.com/endless-sky/endless-sky/pull/4502#issuecomment-526626167).

**Comment by Pointedstick** [comment]
When I try this out, I don't see the hit indication circle appear on my ship's HUD at all.

And in terms of visuals, I think the circle is too thick and the red too bright. It's a bit distracting.

**Comment by Amazinite** [comment]
I'm not really on board with the idea here as a whole. I already know if I'm taking damage based off of my HP dropping. I don't also need a red ring to tell me that I'm currently taking damage. It feels too "FPS-y" to me, like when your character gets shot and you flinch and the screen goes red. 

This exact execution isn't the best either. While the damage indicator doesn't overlap the hull ring, it does obscure it by having such a bright red right next to the yellowish-orange of the hull ring, making it more difficult to tell what hull value I'm actually at while taking damage.

**Pull request closed without merging**